my_map.rb

```diff
@@ -0,0 +1,64 @@
+class Array
+  def my_map
+    array = []
+    if block_given?
+      self.each { |value| array << yield(value) }
+      array
+    else
+      self.to_enum(:map)
+    end
+  end
+  def my_map!
+    if block_given?
+     self.my_map.with_index { |value, i| self[i] = yield(value) }
+    else
+      self.to_enum(:map)
+    end
+  end
+  def my_reduce(sum = nil, symbol = nil)
+    if sum.nil? && symbol.nil?
+      sum = self.first
```

Comment:
Read about `Array.shift`. It will be helpful to make the code more clear.

Reply:
Still not so clear for me. I didn't understand how can I use .shift. Because I don't see array variable in the method. Where should I put .shift?

Reply:
It will be `self.shift` in your case.
```
# self = [1,2,3]
sum = sefl.shift
# self = [2,3]
# sum = 1
```

Reply:
fixed